github/src/main/kotlin/gropius/sync/github/Incoming.kt

```diff
@@ -105,27 +128,119 @@ class Incoming(
      * Sync github to gropius
      */
     suspend fun sync() {
-        val issueGrabber = IssueGrabber(repositoryInfoRepository, mongoOperations)
+        imsConfigManager.findTemplates()
+        for (imsTemplate in imsConfigManager.findTemplates()) {
+            for (ims in imsTemplate.usedIn()) {
+                try {
+                    val imsConfig = IMSConfig(helper, ims)
+                    syncIMS(imsConfig)
+                } catch (e: SyncNotificator.NotificatedError) {
+                    syncNotificator.sendNotification(
+                        ims, SyncNotificator.NotificationDummy(e)
+                    )
+                } catch (e: Exception) {
+                    logger.warn("Error in global sync", e)
+                }
+            }
+        }
+    }
+
+    /**
+     * Request an user token from the auth service
+     * @param ims IMS the token is requested for
+     * @param gropiusUser The user the token should be for
+     * @return token if available
+     */
+    suspend fun getGithubUserToken(ims: IMS, gropiusUser: GropiusUser): String? {
+        return System.getenv("GITHUB_DUMMY_PAT")//TODO: @modellbahnfreak!!
+    }
+
+    /**
+     * Sync one IMS
+     * @param imsConfig the config of the IMS
+     */
+    suspend fun syncIMS(imsConfig: IMSConfig) {
+        val readUser =
+            neoOperations.findById<GropiusUser>(imsConfig.readUser) ?: throw SyncNotificator.NotificatedError(
+                "SYNC_GITHUB_USER_NOT_FOUND"
+            )
+        val token = getGithubUserToken(imsConfig.ims, readUser) ?: throw SyncNotificator.NotificatedError(
+            "SYNC_GITHUB_USER_NO_TOKEN"
+        )
+        val apolloClient = ApolloClient.Builder().serverUrl("https://api.github.com/graphql")
+            .addHttpHeader("Authorization", "bearer $token").build()
+        for (project in imsConfig.ims.projects()) {
+            try {
+                val imsProjectConfig = IMSProjectConfig(helper, imsConfig, project)
+                syncProject(imsProjectConfig, apolloClient)
+            } catch (e: SyncNotificator.NotificatedError) {
+                syncNotificator.sendNotification(
+                    project, SyncNotificator.NotificationDummy(e)
+                )
+            } catch (e: Exception) {
+                logger.warn("Error in IMS sync", e)
+            }
+        }
+    }
+
+    /**
+     * Sync one IMSProject
+     * @param imsProjectConfig the config of the IMSProject
+     * @param apolloClient the client to use4 for grpahql queries
+     */
+    suspend fun syncProject(imsProjectConfig: IMSProjectConfig, apolloClient: ApolloClient) {
```

Comment:
this one seem to be a little bit long

Reply:
97e7d2bad00bc5df055901a6122ebc81ba380105